rust/relay/src/sockets.rs

```diff
@@ -0,0 +1,279 @@
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Received<'b>, Error>> {
+        loop {
+            if let Some(current) = self.current_ready_socket {
+                if let Some(socket) = self.inner.get(&current) {
+                    let (num_bytes, from) = match socket.recv_from(buf) {
+                        Ok(ok) => ok,
+                        Err(e) if e.kind() == io::ErrorKind::WouldBlock => {
+                            self.current_ready_socket = None;
+                            continue;
+                        }
+                        Err(e) => {
+                            self.current_ready_socket = None;
+                            return Poll::Ready(Err(Error::Io(e)));
+                        }
+                    };
+
+                    let (port, _) = token_to_port_and_address_family(current);
+
+                    return Poll::Ready(Ok(Received {
+                        port,
+                        from,
+                        packet: &buf[..num_bytes],
+                    }));
+                }
+            }
+
+            match ready!(self.event_rx.poll_recv(cx)) {
+                Some(Event::NewSocket(token, socket)) => {
+                    self.inner.insert(token, socket);
+                    continue;
+                }
+                Some(Event::SocketReady(ready)) => {
+                    self.current_ready_socket = Some(ready);
+                    continue;
+                }
+                Some(Event::Crashed(error)) => {
+                    return Poll::Ready(Err(Error::MioTaskCrashed(error)));
+                }
+                None => {
+                    panic!("must not poll `Sockets` after mio task exited")
+                }
+            };
+        }
+    }
+}
+
+/// A packet read from a socket.
+#[derive(Debug)]
+pub struct Received<'a> {
+    pub port: u16,
+    pub from: SocketAddr,
+    pub packet: &'a [u8],
+}
+
+#[derive(Debug)]
+pub enum Error {
+    Io(io::Error),
+    MioTaskCrashed(anyhow::Error),
+}
+
+enum Command {
+    NewSocket((u16, AddressFamily)),
+    DisposeSocket(mio::net::UdpSocket),
+}
+
+enum Event {
+    NewSocket(mio::Token, mio::net::UdpSocket),
+    SocketReady(mio::Token),
+    Crashed(anyhow::Error),
+}
+
+fn not_connected(port: u16, address_family: AddressFamily) -> io::Error {
+    io::Error::new(
+        io::ErrorKind::NotConnected,
+        format!("No socket for port {port} on address family {address_family}"),
+    )
+}
+
+/// The [`mio`] worker task which checks for read-readiness on any of our sockets.
+///
+/// This task is connected with the main eventloop via two channels.
+fn mio_worker_task(
+    event_tx: mpsc::Sender<Event>,
+    mut cmd_rx: mpsc::Receiver<Command>,
+) -> Result<()> {
+    let mut poll = mio::Poll::new()?;
+    let mut events = mio::Events::with_capacity(1024);
+
+    loop {
+        poll.poll(&mut events, Some(Duration::from_secs(1)))?; // Suspend for up to 1 second to wait for IO events.
```

Comment:
Would it wake up if there are events before 1 second is elapsed or it will sleep for 1 second no matter what happens?

Reply:
Yeah I think this might need some tuning. I actually didn't think of the delay right after `NewSocket` so perhaps this is better at a lower value? I didn't want this thread to burn through CPU cycles if it is waking up every 50ms or something like that.